src/v/datalake/coordinator/coordinator.cc

```diff
@@ -0,0 +1,149 @@
+ */
+#include "datalake/coordinator/coordinator.h"
+
+#include "base/vlog.h"
+#include "container/fragmented_vector.h"
+#include "datalake/coordinator/state_update.h"
+#include "datalake/logger.h"
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+#include "storage/record_batch_builder.h"
+
+namespace datalake::coordinator {
+
+namespace {
+coordinator::errc convert_stm_errc(coordinator_stm::errc e) {
+    switch (e) {
+    case coordinator_stm::errc::not_leader:
+        return coordinator::errc::not_leader;
+    case coordinator_stm::errc::shutting_down:
+        return coordinator::errc::shutting_down;
+    case coordinator_stm::errc::apply_error:
+        return coordinator::errc::stm_apply_error;
+    case coordinator_stm::errc::replication_error:
+        return coordinator::errc::timedout;
+    }
+}
+} // namespace
+
+std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
+    switch (e) {
+    case coordinator::errc::not_leader:
+        return o << "coordinator::errc::not_leader";
+    case coordinator::errc::shutting_down:
+        return o << "coordinator::errc::shutting_down";
+    case coordinator::errc::stm_apply_error:
+        return o << "coordinator::errc::stm_apply_error";
+    case coordinator::errc::timedout:
+        return o << "coordinator::errc::timedout";
+    }
+}
+
+ss::future<> coordinator::stop_and_wait() {
+    as_.request_abort();
+    return gate_.close();
+}
+
+checked<ss::gate::holder, coordinator::errc> coordinator::maybe_gate() {
+    ss::gate::holder h;
+    if (as_.abort_requested() || gate_.is_closed()) {
+        return errc::shutting_down;
+    }
+    return gate_.hold();
+}
+
+ss::future<checked<std::nullopt_t, coordinator::errc>>
+coordinator::sync_add_files(
+  model::topic_partition tp, chunked_vector<translated_offset_range> entries) {
+    if (entries.empty()) {
+        vlog(datalake_log.debug, "Empty entry requested {}", tp);
+        co_return std::nullopt;
+    }
+    auto gate = maybe_gate();
+    if (gate.has_error()) {
+        co_return gate.error();
+    }
+    vlog(
+      datalake_log.debug,
+      "Sync add files requested {}: [{}, {}], {} files",
+      tp,
+      entries.begin()->start_offset,
+      entries.back().last_offset,
+      entries.size());
+    auto sync_res = co_await stm_.sync(30s);
+    if (sync_res.has_error()) {
+        co_return convert_stm_errc(sync_res.error());
+    }
+    auto added_last_offset = entries.back().last_offset;
+    auto update_res = add_files_update::build(
+      stm_.state(), tp, std::move(entries));
+    if (update_res.has_error()) {
+        // NOTE: rejection here is just an optimization -- the operation would
+        // fail to be applied to the STM anyway.
+        vlog(
+          datalake_log.debug,
+          "Rejecting request to add files for {}: {}",
+          tp,
+          update_res.error());
+        co_return errc::stm_apply_error;
```

Comment:
nit: I think this is a validation error? (since its pre replication).

Reply:
From the caller's perspective they should be the same thing, no? the request does not apply cleanly to the stm

Reply:
I was thinking from a debugging perspective, if the request was rejected due to a race or if was legit validation, nbd, feel free to keep it as is.